### PR TITLE
Remove all unreferenced definitions.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/powdr-labs/powdr/issues/966

### CODE REVIEW

File: ast/src/analyzed/mod.rs [hunk 1]
```diff
@@ -296,6 +190,72 @@ impl<T> Analyzed<T> {
         })
     }
 
+    /// Removes the given definitions and itermediate columns by name. Those must not be referenced
+    /// by any remaining definitions, identities or public declarations.
```
Comment (leonardoalt) [inline ast/src/analyzed/mod.rs:194]: Is this tested after or assumed?

Comment (chriseth) [reply ast/src/analyzed/mod.rs:194]: assumed

File: pipeline/tests/pil.rs [hunk 2]
```diff
@@ -173,13 +173,6 @@ fn test_external_witgen_fails_on_conflicting_external_witness() {
     verify_test_file(f, Default::default(), external_witness).unwrap();
 }
 
-#[test]
-fn test_global() {
-    verify_pil("pil/global.pil", Default::default());
-    // Halo2 would take too long for this.
-    // Starky requires at least one witness column, this test has none.
-}
-
```
Comment (leonardoalt) [inline pipeline/tests/pil.rs:182]: why was this removed?

Comment (chriseth) [reply pipeline/tests/pil.rs:182]: Because the new optimizer removes everything in that file causing an error with pil-verify. The file also does not contain anything meaningful.

File: ast/src/parsed/visitor.rs [hunk 3]
```diff
@@ -14,6 +14,12 @@ pub trait Children<O> {
     fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut O> + '_>;
 }
 
+pub trait AllChildren<O> {
```
Comment (chriseth) [inline ast/src/parsed/visitor.rs:17]: Need to wait for #1199 to be merged.

File: pilopt/src/lib.rs [hunk 4]
```diff
@@ -440,20 +589,6 @@ mod test {
         assert_eq!(optimized, expectation);
     }
 
-    #[test]
```
Comment (chriseth) [inline pilopt/src/lib.rs:443]: Acutally I think there is a way to make this test work.

File: ast/src/analyzed/mod.rs [hunk 5]
```diff
@@ -210,7 +210,8 @@ impl<T> Analyzed<T> {
 
         let mut handle_symbol = |new_id: u64, symbol: &Symbol| -> u64 {
             let length = symbol.length.unwrap_or(1);
-            for i in 0..length {
+            // Empty arrays still need ID replacement
+            for i in 0..(if length == 0 { 1 } else { length }) {
```
Comment (leonardoalt) [inline ast/src/analyzed/mod.rs:214]: `max(1, length)`?

**Review by leonardoalt — APPROVED** [review]
